Sources/PackageLoading/TargetSourcesBuilder.swift

```diff
@@ -0,0 +1,311 @@
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import Foundation
+import TSCBasic
+import PackageModel
+import TSCUtility
+
+/// An utility to compute the source/resource files of a target.
+public struct TargetSourcesBuilder {
+    /// The target for which we're computing source/resource files.
+    public let target: TargetDescription
+
+    /// The engine for emitting diagnostics.
+    let diags: DiagnosticsEngine
+
+    /// The path of the target.
+    public let targetPath: AbsolutePath
+
+    /// The rules that can be applied to files in the target.
+    public let rules: [FileRuleDescription]
+
+    /// The tools version associated with the target's package.
+    public let toolsVersion: ToolsVersion
+
+    /// The set of paths that should be excluded from any consideration.
+    public let excludedPaths: Set<AbsolutePath>
+
+    /// The file system to operate on.
+    public let fs: FileSystem
+
+    /// Create a new target builder.
+    public init(
+        target: TargetDescription,
+        path: AbsolutePath,
+        rules: [FileRuleDescription] = [],
+        toolsVersion: ToolsVersion = .currentToolsVersion,
+        fs: FileSystem = localFileSystem,
+        diags: DiagnosticsEngine
+    ) {
+        self.target = target
+        self.diags = diags
+        self.targetPath = path
+        self.rules = FileRuleDescription.builtinRules + rules
+        self.toolsVersion = toolsVersion
+        self.fs = fs
+        let excludedPaths = target.exclude.map{ path.appending(RelativePath($0)) }
+        self.excludedPaths = Set(excludedPaths)
+
+      #if DEBUG
+        validateRules(self.rules)
+      #endif
+    }
+
+    /// Emits an error in debug mode if we have conflicting rules for any file type.
+    private func validateRules(_ rules: [FileRuleDescription]) {
+        var extensionMap: [String: FileRuleDescription] = [:]
+        for rule in rules {
+            for ext in rule.fileTypes {
+                if let existingRule = extensionMap[ext] {
+                    diags.emit(.error("conflicting rules \(rule) and \(existingRule) for extension \(ext)"))
+                }
+                extensionMap[ext] = rule
+            }
+        }
+    }
+
+    /// Run the builder to produce the sources of the target.
+    public func run() -> (sources: Sources, resources: [Resource]) {
+        let contents = computeContents()
+        var pathToRule: [AbsolutePath: FileRuleDescription.Rule] = [:]
+
+        for path in contents {
+            pathToRule[path] = findRule(for: path)
+        }
+
+        // Emit an error if we found files without a matching rule in tools version >= 5.2
+        if toolsVersion >= .v5_2 {
+            let filesWithNoRules = pathToRule.filter{ $0.value == .none }
+            if !filesWithNoRules.isEmpty {
+                var error = "found \(filesWithNoRules.count) file(s) without any matching rule; explicitly declare them as resources or exclude from the target\n"
```

Comment:
I think we shouldn't use the term "rule" in this error message, because rules will be an implementation detail for now. How about something like "found X file(s) which are unhandled, explicitly..."

Reply:
yeah